File: linux/test-app-local-dependency/src/HttpsTab.cpp
/*
 * Copyright (c) 2022 Taner Sener
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include "HttpsTab.h"
#include "Constants.h"
#include "Popup.h"
#include <FFmpegKitConfig.h>
#include <cstdlib>
#define RAPIDJSON_ASSERT(x)
#include "rapidjson/writer.h"
#include "rapidjson/stringbuffer.h"

using namespace ffmpegkit;

static std::recursive_mutex outputMutex;

std::string toString(const rapidjson::Value& value) {
    rapidjson::StringBuffer buffer;
    rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
    value.Accept(writer);
    return std::string(buffer.GetString(), buffer.GetSize());
}

static gboolean appendLog(const std::pair<ffmpegkittest::HttpsTab*,const std::string>* parameters) {
    ffmpegkittest::HttpsTab* httpsTab = parameters->first;
    auto string = parameters->second;
    httpsTab->appendOutput(string);
    delete parameters;
    return FALSE;
}

void appendLogToMainLoop(const ffmpegkittest::HttpsTab* httpsTab,const std::string string) {
    g_idle_add((GSourceFunc)appendLog, new std::pair<const ffmpegkittest::HttpsTab*,const std::string>(httpsTab, string));
}

ffmpegkittest::HttpsTab::HttpsTab() : parentWindow(nullptr) {
    urlText.set_placeholder_text("Enter https url");
    Util::applyEditTextStyle(urlText);

    getInfoFromUrlButton.set_label("GET INFO FROM URL");
    getInfoFromUrlButton.set_size_request(120, 30);
    getInfoFromUrlButton.set_tooltip_text(Constants::HttpsTestTooltipText);
    getInfoFromUrlButton.signal_clicked().connect(sigc::bind(sigc::mem_fun(*this, &HttpsTab::runGetMediaInformation), 1));
    Util::applyButtonStyle(getInfoFromUrlButton);
    getInfoFromUrlButtonBox.pack_start(getInfoFromUrlButton, Gtk::PACK_EXPAND_PADDING);

    getRandomInfoButton1.set_label("GET RANDOM INFO");
    getRandomInfoButton1.set_size_request(120, 30);
    getRandomInfoButton1.signal_clicked().connect(sigc::bind(sigc::mem_fun(*this, &HttpsTab::runGetMediaInformation), 2));
    Util::applyButtonStyle(getRandomInfoButton1);
    getRandomInfoButton1Box.pack_start(getRandomInfoButton1, Gtk::PACK_EXPAND_PADDING);

    getRandomInfoButton2.set_label("GET RANDOM INFO");
    getRandomInfoButton2.set_size_request(120, 30);
    getRandomInfoButton2.signal_clicked().connect(sigc::bind(sigc::mem_fun(*this, &HttpsTab::runGetMediaInformation), 3));
    Util::applyButtonStyle(getRandomInfoButton2);
    getRandomInfoButton2Box.pack_start(getRandomInfoButton2, Gtk::PACK_EXPAND_PADDING);

    getInfoAndFailButton.set_label("GET INFO AND FAIL");
    getInfoAndFailButton.set_size_request(120, 30);

    getInfoAndFailButton.signal_clicked().connect(sigc::bind(sigc::mem_fun(*this, &HttpsTab::runGetMediaInformation), 4));
    Util::applyButtonStyle(getInfoAndFailButton);
    getInfoAndFailButtonBox.pack_start(getInfoAndFailButton, Gtk::PACK_EXPAND_PADDING);

    outputText.set_editable(false);
    Util::applyOutputTextStyle(outputText);
    outputTextWindow.add(outputText);

    pack_start(urlText, Gtk::PACK_SHRINK);
    pack_start(getInfoFromUrlButtonBox, Gtk::PACK_SHRINK);
    pack_start(getRandomInfoButton1Box, Gtk::PACK_SHRINK);
    pack_start(getRandomInfoButton2Box, Gtk::PACK_SHRINK);
    pack_start(getInfoAndFailButtonBox, Gtk::PACK_SHRINK);
    add(outputTextWindow);
}

void ffmpegkittest::HttpsTab::setActive() {
    std::cout << "Https Tab Activated" << std::endl;
    FFmpegKitConfig::enableLogCallback(nullptr);
    FFmpegKitConfig::enableStatisticsCallback(nullptr);
}

void ffmpegkittest::HttpsTab::setParentWindow(Gtk::Window* parentWindow) {
    this->parentWindow = parentWindow;
}

void ffmpegkittest::HttpsTab::appendOutput(const std::string& string) {
    outputText.get_buffer()->set_text(outputText.get_buffer()->get_text() + string);
    Glib::RefPtr<Gtk::Adjustment> adj = outputText.get_vadjustment();
    adj->set_value(adj->get_upper());
}

void ffmpegkittest::HttpsTab::clearOutput() {
    outputText.get_buffer()->set_text("");
}

void ffmpegkittest::HttpsTab::runGetMediaInformation(const int buttonNumber) {

    // SELECT TEST URL
    std::string testUrl;
    switch (buttonNumber) {
        case 1: {
            testUrl = urlText.get_text();
            if (testUrl.empty()) {
                testUrl = HttpsTestDefaultUrl;
                urlText.set_text(testUrl);
            }
        }
        break;
        case 2:
        case 3: {
            testUrl = getRandomTestUrl();
        }
        break;
        case 4:
        default: {
            testUrl = HttpsTestFailUrl;
            urlText.set_text(testUrl);
        }
    }

    std::cout << "Testing HTTPS with for button " << buttonNumber << " using url " << testUrl << "." << std::endl;

    if (buttonNumber == 4) {

        // ONLY THIS BUTTON CLEARS THE TEXT VIEW
        clearOutput();
    }

    // EXECUTE
    FFprobeKit::getMediaInformationAsync(testUrl, createNewCompleteCallback());

}

std::string ffmpegkittest::HttpsTab::getRandomTestUrl() {
    switch (std::rand() % 3) {
        case 0:
            return HttpsTestRandomUrl1;
        case 1:
            return HttpsTestRandomUrl2;
        default:
            return HttpsTestRandomUrl3;
    }
}

MediaInformationSessionCompleteCallback ffmpegkittest::HttpsTab::createNewCompleteCallback() {
    return [this](auto session) {
        std::unique_lock<std::recursive_mutex> lock(outputMutex);
        auto information = session->getMediaInformation();
        if (information == nullptr) {
            appendLogToMainLoop(this, "Get media information failed\n");
            appendLogToMainLoop(this, "State: " + FFmpegKitConfig::sessionStateToString(session->getState()) + "\n");
            appendLogToMainLoop(this, "Duration: " + std::to_string(session->getDuration()) + "\n");
            if (session->getReturnCode() != nullptr) {
                appendLogToMainLoop(this, "Return Code: " + std::to_string(session->getReturnCode()->getValue()) + "\n");
            }
            appendLogToMainLoop(this, "Fail stack trace: " + session->getFailStackTrace() + "\n");
            appendLogToMainLoop(this, "Output: " + session->getOutput() + "\n");
        } else {
            if (information->getFilename() != nullptr) {
                appendLogToMainLoop(this, "Media information for " + *information->getFilename() + "\n");
            }
            if (information->getFormat() != nullptr) {
                appendLogToMainLoop(this, "Format: " + *information->getFormat() + "\n");
            }
            if (information->getBitrate() != nullptr) {
                appendLogToMainLoop(this, "Bitrate: " + *information->getBitrate() + "\n");
            }
            if (information->getDuration() != nullptr) {
                appendLogToMainLoop(this, "Duration: " + *information->getDuration() + "\n");
            }
            if (information->getStartTime() != nullptr) {
                appendLogToMainLoop(this, "Start time: " + *information->getStartTime() + "\n");
            }
            if (information->getTags() != nullptr) {
                auto tags = information->getTags();
                for (auto tagIterator = tags->MemberBegin(); tagIterator != tags->MemberEnd(); ++tagIterator) {
                    const char* tagName = tagIterator->name.GetString();
                    appendLogToMainLoop(this, std::string("Tag: ") + tagName + ":" + toString(tagIterator->value) + "\n");
                }
            }
            if (information->getStreams() != nullptr) {
                auto streams = information->getStreams();
                std::for_each(streams->cbegin(), streams->cend(), [this](const std::shared_ptr<ffmpegkit::StreamInformation>& stream) {
                    if (stream->getIndex() != nullptr) {
                        appendLogToMainLoop(this, "Stream index: " + std::to_string(*stream->getIndex()) + "\n");
                    }
                    if (stream->getType() != nullptr) {
                        appendLogToMainLoop(this, "Stream type: " + *stream->getType() + "\n");
                    }
                    if (stream->getCodec() != nullptr) {
                        appendLogToMainLoop(this, "Stream codec: " + *stream->getCodec() + "\n");
                    }
                    if (stream->getCodecLong() != nullptr) {
                        appendLogToMainLoop(this, "Stream codec long: " + *stream->getCodecLong() + "\n");
                    }
                    if (stream->getFormat() != nullptr) {
                        appendLogToMainLoop(this, "Stream format: " + *stream->getFormat() + "\n");
                    }

                    if (stream->getWidth() != nullptr) {
                        appendLogToMainLoop(this, "Stream width: " + std::to_string(*stream->getWidth()) + "\n");
                    }
                    if (stream->getHeight() != nullptr) {
                        appendLogToMainLoop(this, "Stream height: " + std::to_string(*stream->getHeight()) + "\n");
                    }

                    if (stream->getBitrate() != nullptr) {
                        appendLogToMainLoop(this, "Stream bitrate: " + *stream->getBitrate() + "\n");
                    }
                    if (stream->getSampleRate() != nullptr) {
                        appendLogToMainLoop(this, "Stream sample rate: " + *stream->getSampleRate() + "\n");
                    }
                    if (stream->getSampleFormat() != nullptr) {
                        appendLogToMainLoop(this, "Stream sample format: " + *stream->getSampleFormat() + "\n");
                    }
                    if (stream->getChannelLayout() != nullptr) {
                        appendLogToMainLoop(this, "Stream channel layout: " + *stream->getChannelLayout() + "\n");
                    }

                    if (stream->getSampleAspectRatio() != nullptr) {
                        appendLogToMainLoop(this, "Stream sample aspect ratio: " + *stream->getSampleAspectRatio() + "\n");
                    }
                    if (stream->getDisplayAspectRatio() != nullptr) {
                        appendLogToMainLoop(this, "Stream display ascpect ratio: " + *stream->getDisplayAspectRatio() + "\n");
                    }
                    if (stream->getAverageFrameRate() != nullptr) {
                        appendLogToMainLoop(this, "Stream average frame rate: " + *stream->getAverageFrameRate() + "\n");
                    }
                    if (stream->getRealFrameRate() != nullptr) {
                        appendLogToMainLoop(this, "Stream real frame rate: " + *stream->getRealFrameRate() + "\n");
                    }
                    if (stream->getTimeBase() != nullptr) {
                        appendLogToMainLoop(this, "Stream time base: " + *stream->getTimeBase() + "\n");
                    }
                    if (stream->getCodecTimeBase() != nullptr) {
                        appendLogToMainLoop(this, "Stream codec time base: " + *stream->getCodecTimeBase() + "\n");
                    }

                    if (stream->getTags() != nullptr) {
                        auto tags = stream->getTags();
                        for (auto tagIterator = tags->MemberBegin(); tagIterator != tags->MemberEnd(); ++tagIterator) {
                            const char* tagName = tagIterator->name.GetString();
                            appendLogToMainLoop(this, std::string("Stream tag: ") + tagName + ":" + toString(tagIterator->value) + "\n");
                        }
                    }
                });
            }

            if (information->getChapters() != nullptr) {
                auto chapters = information->getChapters();
                std::for_each(chapters->cbegin(), chapters->cend(), [this](const std::shared_ptr<ffmpegkit::Chapter>& chapter) {
                    if (chapter->getId() != nullptr) {
                        appendLogToMainLoop(this, "Chapter id: " + std::to_string(*chapter->getId()) + "\n");
                    }
                    if (chapter->getTimeBase() != nullptr) {
                        appendLogToMainLoop(this, "Chapter time base: " + *chapter->getTimeBase() + "\n");
                    }
                    if (chapter->getStart() != nullptr) {
                        appendLogToMainLoop(this, "Chapter start: " + std::to_string(*chapter->getStart()) + "\n");
                    }
                    if (chapter->getStartTime() != nullptr) {
                        appendLogToMainLoop(this, "Chapter start time: " + *chapter->getStartTime() + "\n");
                    }
                    if (chapter->getEnd() != nullptr) {
                        appendLogToMainLoop(this, "Chapter end: " + std::to_string(*chapter->getEnd()) + "\n");
                    }
                    if (chapter->getEndTime() != nullptr) {
                        appendLogToMainLoop(this, "Chapter end time: " + *chapter->getEndTime() + "\n");
                    }
                    if (chapter->getTags() != nullptr) {
                        auto tags = chapter->getTags();
                        for (auto tagIterator = tags->MemberBegin(); tagIterator != tags->MemberEnd(); ++tagIterator) {
                            const char* tagName = tagIterator->name.GetString();
                            appendLogToMainLoop(this, std::string("Chapter tag: ") + tagName + ":" + toString(tagIterator->value) + "\n");
                        }
                    }
                });
            }
        }
    };
}

File: flutter/test-app-local-dependency/lib/https_tab.dart
/*
 * Copyright (c) 2018-2022 Taner Sener
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import 'dart:math';

import 'package:ffmpeg_kit_flutter/chapter.dart';
import 'package:ffmpeg_kit_flutter/ffmpeg_kit_config.dart';
import 'package:ffmpeg_kit_flutter/ffprobe_kit.dart';
import 'package:ffmpeg_kit_flutter/media_information_session.dart';
import 'package:ffmpeg_kit_flutter/stream_information.dart';
import 'package:flutter/material.dart';

import 'abstract.dart';
import 'popup.dart';
import 'tooltip.dart';
import 'util.dart';

class HttpsTab {
  static const String HTTPS_TEST_DEFAULT_URL =
      "https://download.blender.org/peach/trailer/trailer_1080p.ogg";

  static const String HTTPS_TEST_FAIL_URL =
      "https://download2.blender.org/peach/trailer/trailer_1080p.ogg";

  static const String HTTPS_TEST_RANDOM_URL_1 =
      "https://filesamples.com/samples/video/mov/sample_640x360.mov";

  static const String HTTPS_TEST_RANDOM_URL_2 =
      "https://filesamples.com/samples/audio/mp3/sample3.mp3";

  static const String HTTPS_TEST_RANDOM_URL_3 =
      "https://filesamples.com/samples/image/webp/sample1.webp";

  static final Random testUrlRandom = new Random();

  late Refreshable _refreshable;
  late TextEditingController _urlText;
  String _outputText = "";

  void init(Refreshable refreshable) {
    _refreshable = refreshable;
    _urlText = TextEditingController();
    this.clearOutput();
  }

  void setActive() {
    print("Https Tab Activated");
    FFmpegKitConfig.enableLogCallback(null);
    FFmpegKitConfig.enableStatisticsCallback(null);
    showPopup(HTTPS_TEST_TOOLTIP_TEXT);
  }

  void appendOutput(String logMessage) {
    _outputText += logMessage;
    _refreshable.refresh();
  }

  void clearOutput() {
    _outputText = "";
    _refreshable.refresh();
  }

  String getRandomTestUrl() {
    switch (testUrlRandom.nextInt(3)) {
      case 0:
        return HTTPS_TEST_RANDOM_URL_1;
      case 1:
        return HTTPS_TEST_RANDOM_URL_2;
      default:
        return HTTPS_TEST_RANDOM_URL_3;
    }
  }

  void runGetMediaInformation(int buttonNumber) {
    // SELECT TEST URL
    String testUrl = "";
    switch (buttonNumber) {
      case 1:
        {
          testUrl = _urlText.text;
          if (testUrl.trim().length <= 0) {
            testUrl = HTTPS_TEST_DEFAULT_URL;
            _urlText.text = testUrl;
          }
          break;
        }
      case 2:
      case 3:
        {
          testUrl = this.getRandomTestUrl();
          break;
        }
      case 4:
      default:
        {
          testUrl = HTTPS_TEST_FAIL_URL;
          _urlText.text = testUrl;
        }
    }

    ffprint(
        "Testing HTTPS with for button ${buttonNumber} using url ${testUrl}.");

    if (buttonNumber == 4) {
      // ONLY THIS BUTTON CLEARS THE TEXT VIEW
      this.clearOutput();
    }

    // EXECUTE
    FFprobeKit.getMediaInformation(testUrl)
        .then((MediaInformationSession session) async {
      var information = await session.getMediaInformation();

      if (information == null) {
        final state =
            FFmpegKitConfig.sessionStateToString(await session.getState());
        final returnCode = await session.getReturnCode();
        final failStackTrace = await session.getFailStackTrace();
        final duration = await session.getDuration();
        final output = await session.getOutput();

        this.appendOutput("Get media information failed\n");
        this.appendOutput("State: ${state}\n");
        this.appendOutput("Duration: ${duration}\n");
        this.appendOutput("Return Code: ${returnCode}\n");
        this.appendOutput(
            "Fail stack trace: ${notNull(failStackTrace, "\\n")}\n");
        this.appendOutput("Output: ${output}\n");
      } else {
        this.appendOutput(
            "Media information for ${information.getFilename()}\n");

        if (information.getFormat() != null) {
          this.appendOutput("Format: ${information.getFormat()}\n");
        }
        if (information.getBitrate() != null) {
          this.appendOutput("Bitrate: ${information.getBitrate()}\n");
        }
        if (information.getDuration() != null) {
          this.appendOutput("Duration: ${information.getDuration()}\n");
        }
        if (information.getStartTime() != null) {
          this.appendOutput("Start time: ${information.getStartTime()}\n");
        }
        if (information.getTags() != null) {
          final tags = information.getTags();
          if (tags != null) {
            tags.forEach((key, value) {
              this.appendOutput("Tag: ${key}:${tags[key]}\n");
            });
          }
        }

        List<StreamInformation> streams = information.getStreams();
        for (var i = 0; i < streams.length; ++i) {
          StreamInformation stream = streams[i];
          if (stream.getIndex() != null) {
            this.appendOutput("Stream index: ${stream.getIndex()}\n");
          }
          if (stream.getType() != null) {
            this.appendOutput("Stream type: ${stream.getType()}\n");
          }
          if (stream.getCodec() != null) {
            this.appendOutput("Stream codec: ${stream.getCodec()}\n");
          }
          if (stream.getCodecLong() != null) {
            this.appendOutput("Stream codec long: ${stream.getCodecLong()}\n");
          }
          if (stream.getFormat() != null) {
            this.appendOutput("Stream format: ${stream.getFormat()}\n");
          }
          if (stream.getWidth() != null) {
            this.appendOutput("Stream width: ${stream.getWidth()}\n");
          }
          if (stream.getHeight() != null) {
            this.appendOutput("Stream height: ${stream.getHeight()}\n");
          }
          if (stream.getBitrate() != null) {
            this.appendOutput("Stream bitrate: ${stream.getBitrate()}\n");
          }
          if (stream.getSampleRate() != null) {
            this.appendOutput(
                "Stream sample rate: ${stream.getSampleRate()}\n");
          }
          if (stream.getSampleFormat() != null) {
            this.appendOutput(
                "Stream sample format: ${stream.getSampleFormat()}\n");
          }
          if (stream.getChannelLayout() != null) {
            this.appendOutput(
                "Stream channel layout: ${stream.getChannelLayout()}\n");
          }
          if (stream.getSampleAspectRatio() != null) {
            this.appendOutput(
                "Stream sample aspect ratio: ${stream.getSampleAspectRatio()}\n");
          }
          if (stream.getDisplayAspectRatio() != null) {
            this.appendOutput(
                "Stream display ascpect ratio: ${stream.getDisplayAspectRatio()}\n");
          }
          if (stream.getAverageFrameRate() != null) {
            this.appendOutput(
                "Stream average frame rate: ${stream.getAverageFrameRate()}\n");
          }
          if (stream.getRealFrameRate() != null) {
            this.appendOutput(
                "Stream real frame rate: ${stream.getRealFrameRate()}\n");
          }
          if (stream.getTimeBase() != null) {
            this.appendOutput("Stream time base: ${stream.getTimeBase()}\n");
          }
          if (stream.getCodecTimeBase() != null) {
            this.appendOutput(
                "Stream codec time base: ${stream.getCodecTimeBase()}\n");
          }
          if (stream.getTags() != null) {
            final tags = stream.getTags();
            if (tags != null) {
              tags.forEach((key, value) {
                this.appendOutput("Stream tag: ${key}:${tags[key]}\n");
              });
            }
          }
        }

        List<Chapter> chapters = information.getChapters();
        for (var i = 0; i < chapters.length; ++i) {
          Chapter chapter = chapters[i];
          if (chapter.getId() != null) {
            appendOutput("Chapter id: ${chapter.getId()}\n");
          }
          if (chapter.getTimeBase() != null) {
            appendOutput("Chapter time base: ${chapter.getTimeBase()}\n");
          }
          if (chapter.getStart() != null) {
            appendOutput("Chapter start: ${chapter.getStart()}\n");
          }
          if (chapter.getStartTime() != null) {
            appendOutput("Chapter start time: ${chapter.getStartTime()}\n");
          }
          if (chapter.getEnd() != null) {
            appendOutput("Chapter end: ${chapter.getEnd()}\n");
          }
          if (chapter.getEndTime() != null) {
            appendOutput("Chapter end time: ${chapter.getEndTime()}\n");
          }
          if (chapter.getTags() != null) {
            final tags = chapter.getTags();
            if (tags != null) {
              tags.forEach((key, value) {
                this.appendOutput("Chapter tag: ${key}:${tags[key]}\n");
              });
            }
          }
        }
      }
    });
  }

  String getOutputText() => _outputText;

  TextEditingController getUrlText() => _urlText;

  void dispose() {
    _urlText.dispose();
  }
}
